api/genericresource/resource_management.go

```diff
@@ -0,0 +1,62 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assigns GenericResources to the task by taking them from the node's GenericResource
+// and storing them in the taken store
+func Claim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+	t, err := getGenericResourceDiscrete(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		if nr.Discrete < t.Discrete {
+			return NewError("cannot provide enough of scalar resource requested %d, available %d",
+				t.Discrete, nr.Discrete)
+		}
+		nr.Discrete -= t.Discrete
+		taken.Resource = NewDiscrete(t.Discrete)
+	case *api.GenericResource_Set:
+		i := uint64(0)
+		res := NewSet()
+
+		for k := range nr.Set.Val {
+			if i >= t.Discrete {
```

Comment:
> If we allow generic resource of a Set {blue, yellow}, I think we should allow a service to claim specific element, like all the tasks should claim yellow

Let's take a specific example:
- node 1: disks = {sda0, sda1, sda2}
- node 2: disks = {sda0, sda1}

Suppose you want to take specifically sda0 for all your tasks. Then maybe you don't have the right structure and need to add another resource ?

> I don't think that's right. We may not implement Set in the first draft if we do not find use case.

Well the use case is described in the design doc :)
The most important one is exclusive access (ie: GPUs, FPGAs, ..., basically every kind of hardware resource) as well as specific resource assignement :)